src/event.c

```diff
@@ -346,6 +443,23 @@ bool event_loop(void) {
 
 		unsigned int curgen = io_tree.generation;
 
+
+#ifdef HAVE_SYS_EPOLL_H
+
+		for(int i = 0; i < n; i++) {
+			io_t *io = events[i].data.ptr;
+
+			if(events[i].events & EPOLLOUT && io->flags & IO_WRITE) {
+				io->cb(io->data, IO_WRITE);
+			}
+
+			if(events[i].events & EPOLLIN && curgen == io_tree.generation && io->flags & IO_READ) {
+				io->cb(io->data, IO_READ);
+			}
+		}
```

Comment:
Yes, but we also need an `if(curgen != io_tre.generation) break;` at the end of the loop, because a reading data from one fd might cause data to be sent to another fd in the callback, and if that sending fails it might close the connection, and hence invalidate the `io_t` pointer of a subsequent event in the loop.

Reply:
Done